src/components/number-input/number-input.tsx

```diff
@@ -0,0 +1,253 @@
+        }
+    }
+
+    private committed = true;
+
+    private validate(value?: number) {
+        const {min, max} = this.props;
+        return isNumber(value) ?
+          Math.min(max!, Math.max(min!, value))
+          : value;
+    }
+
+    private commit(value?: number) {
+        const {onChange} = this.props;
+        const valueInRange = this.validate(value);
+
+        this.updateValue(valueInRange);
+
+        if (!this.committed) {
+            onChange!(valueInRange);
+            this.committed = true;
+        }
+    }
+
+    private revert() {
+        const {value} = this.props;
+        this.updateValue(value);
+        this.committed = true;
+    }
+
+    private updateValue(next?: number) {
+        const {value} = this.state;
+
+        if(value !== next) {
+            this.committed = false;
+            this.setState({value: next});
+        }
+    }
+
+    private stepValue(e: React.SyntheticEvent<HTMLElement>, direction: Direction) {
+        const {value} = this.state;
+        const {step} = this.props;
+        const next = (direction == Direction.Increase ?
+            isNumber(value) ? value + step! : step! :
+            isNumber(value) ? value - step! : -step!);
+
+        this.commit(next);
+    }
+
+    private handleFocus: React.FocusEventHandler<HTMLElement> =
+        e => this.setState({focus: true});
+
+    private handleBlur: React.FocusEventHandler<HTMLElement> =
+        e => this.setState({focus: false});
+
+    private handleIncrement: React.MouseEventHandler<HTMLElement> =
+        e => this.stepValue(e, Direction.Increase);
+
+    private handleDecrement: React.MouseEventHandler<HTMLElement> =
+        e => this.stepValue(e, Direction.Decrease);
+
+    private handleInputKeyDown: React.KeyboardEventHandler<HTMLElement> =
+        e => {
+            switch (e.keyCode) {
+                case KeyCodes.UP:
+                    this.stepValue(e, Direction.Increase);
+                    e.preventDefault();
+                    break;
+                case KeyCodes.DOWN:
+                    this.stepValue(e, Direction.Decrease);
+                    e.preventDefault();
+                    break;
+                case KeyCodes.ENTER:
+                    this.commit(this.state.value);
+                    e.preventDefault();
+                    break;
+                case KeyCodes.ESCAPE:
+                    this.revert();
+                    e.preventDefault();
+                    break;
+            }
+        }
+
+    private handleInputBlur: React.FocusEventHandler<HTMLInputElement> =
+        e => this.commit(this.state.value);
+
+    private handleInputChange: React.ChangeEventHandler<HTMLInputElement> =
+        e => {
+            const {onInput} = this.props;
+            const value = e.target.value;
+            const next = value !== '' ?
+                Number(e.target.value) :
+                undefined;
+
+            this.updateValue(next);
+            onInput!(value);
```

Comment:
Code style reasons mostly. I generally like to extract all the necessary things at the start of the function and have more logic point-free. I think its more readable this way.